opencog/atoms/proto/ValueFactory.h

```diff
@@ -0,0 +1,98 @@
+#ifndef _VALUE_FACTORY_H_
+#define _VALUE_FACTORY_H_
+
+#include "ProtoAtom.h"
+#include <opencog/atoms/base/Handle.h>
+
+#include <exception>
+#include <map>
+#include <typeinfo>
+#include <typeindex>
+#include <vector>
+
+namespace opencog
+{
+using CreateProto = ProtoAtomPtr (*) (...);
+
+struct FuncRegister
+{
+    CreateProto func;
+    std::vector<std::type_index> args;
+};
+
+class ValueFactory
+{
+private:
+
+#define THROW_ERROR \
+    (throw std::invalid_argument("There was no function found for the given type."))
+
```

Comment:
For historical reasons, we should continue to use the exceptions defined in cogutils.  They have a standard framework for printing, error logging, and the scheme code knows how to catch and rethrow them to the user.